foundry/mouseInteractionManager.d.ts

```diff
@@ -1,213 +1,257 @@
 /**
  * Handle mouse interaction events for a Canvas object.
- *
  * There are three phases of events: hover, click, and drag
+ * @typeParam O - Canvas object this instance handles events for
+ * @typeParam T - Target object for mouseinteraction events. Generally a `ControlIcon` or `O`.
+ *                (default: `O`)
  *
  * Hover Events:
- *
- *      _handleMouseOver
- *              action: hoverIn
- *      _handleMouseOut
- *              action: hoverOut
+ * _handleMouseOver
+ *  action: hoverIn
+ * _handleMouseOut
+ *  action: hoverOut
  *
  * Left Click and Double-Click
- *
- *      _handleMouseDown
- *              action: clickLeft
- *              action: clickLeft2
+ * _handleMouseDown
+ *  action: clickLeft
+ *  action: clickLeft2
  *
  * Right Click and Double-Click
- *
- *      _handleRightDown
- *              action: clickRight
- *              action: clickRight2
+ * _handleRightDown
+ *  action: clickRight
+ *  action: clickRight2
  *
  * Drag and Drop
- *
- *      _handleMouseMove
- *              action: dragLeftStart
- *              action: dragLeftMove
- *              action: dragRightStart
- *              action: dragLeftMove
- *      _handleMouseUp
- *              action: dragLeftDrop
- *              action: dragRightDrop
- *      _handleDragCancel
- *              action: dragLeftCancel
- *              action: dragRightCancel
+ * _handleMouseMove
+ *  action: dragLeftStart
+ *  action: dragLeftMove
+ *  action: dragRightStart
+ *  action: dragLeftMove
+ * _handleMouseUp
+ *  action: dragLeftDrop
+ *  action: dragRightDrop
+ * _handleDragCancel
+ *  action: dragLeftCancel
+ *  action: dragRightCancel
  */
-declare class MouseInteractionManager {
-  callbacks: object;
+declare class MouseInteractionManager<O extends PIXI.Container = PIXI.Container, T extends PIXI.Container = O> {
+  constructor(
+    object: O,
+    layer: MouseInteractionManager['layer'],
+    permissions?: MouseInteractionManager['permissions'],
+    callbacks?: MouseInteractionManager['callbacks'],
+    options?: MouseInteractionManager['options']
+  );
+  object: O;
+  layer: PIXI.Container;
+  permissions: Partial<
+    Record<
+      | 'clickLeft'
+      | 'clickLeft2'
+      | 'clickRight'
+      | 'clickRight2'
+      | 'dragLeftCancel'
+      | 'dragLeftDrop'
+      | 'dragLeftMove'
+      | 'dragLeftStart'
+      | 'dragRightCancel'
+      | 'dragRightDrop'
+      | 'dragRightMove'
+      | 'dragRightStart'
+      | 'hoverIn'
+      | 'hoverOut',
+      ((user: User, event: PIXI.InteractionEvent) => boolean) | boolean
+    >
+  >;
+  callbacks: Record<keyof this['permissions'], ((event: Event | PIXI.InteractionEvent) => any) | null>;
```

Comment:
Does anything speak against using `unknown` as return type for the callbacks?

Reply:
It's an arbitrary distinction.

Reply:
I thnik `unknown` convey "it doesn't matter" while `any` conveys "we don't know". So I would prefer `unknown` but it's not that important in general I think.